File: java-leetcode/src/main/java/dev/codefactory/leetcode/core/common/ListNode.java
package dev.codefactory.leetcode.core.common;


import java.util.LinkedList;
import java.util.List;

public class ListNode {
     public int val;
     public ListNode next;

     public ListNode() {}
     public ListNode(int val) {
         this.val = val;
     }

     public ListNode(int val, ListNode next) {
         this.val = val; this.next = next;
     }

     public static ListNode from(int[] arr) {
         if (arr==null || arr.length<1) return null;

         ListNode head = new ListNode(arr[arr.length-1]);

         for (int i=arr.length-2; i>=0; i--) {
             ListNode newNode = new ListNode(arr[i]);
             newNode.next = head;
             head = newNode;
         }

         return head;
     }

     public int[] toArray() {
         List<Integer> temp = new LinkedList<>();
         temp.add(val);

         ListNode current = next;
         while (current!=null) {
             temp.add(current.val);
             current = current.next;
         }

         return temp.stream().mapToInt(Integer::intValue).toArray();
     }

    @Override
    public String toString() {
        return "" + val;
    }
}

File: kotlin-leetcode/src/main/kotlin/dev/codefactory/leetcode/core/shared/ListNode.kt
package dev.codefactory.leetcode.core.shared

class ListNode(var `val`: Int) {
    var next: ListNode? = null

    override fun toString(): String {
        if (next==null) return "$`val`"
        return "$`val` -> ${next.toString()}"
    }

    override fun equals(other: Any?): Boolean {
        if (other == null || other !is ListNode) {
            return false
        }

        if (next!=null && other.next!=null) {
            return `val` == other.`val` && next!! == other.next
        }

        if (next==null && other.next==null) {
            return `val` == other.`val`
        }

        return false
    }

    override fun hashCode(): Int {
        var result = `val`
        result = 31 * result + (next?.hashCode() ?: 0)
        return result
    }
}